### ISSUE TITLE
Nested mapping & MappingTarget - "cannot find symbol"

### ISSUE THREAD

**Issue by bcole** [mapstruct/mapstruct#1650]
If you have nested mapping (`a.b.c` -> `aPrime.cPrime`) and want to use a `@MappingTarget`, you will run into a Java compilation error.

Consider the following example.

Source classes:
```
public class A {
    private B b;
}

public class B {
    private C c;
}

public class C {}
```

Target classes:
```
public class APrime {
    private CPrime cPrime;
}

public class CPrime {}
```

And Mapper:
```
@Mapper
public interface AMapper {
    @Mapping(source = "b.c", target = "cPrime")
    APrime toAPrime(A a, @MappingTarget APrime mappingTarget);

    CPrime toCPrime(C c, @MappingTarget CPrime mappingTarget);
}
```

Trying to generate the mapper implementation will cause a Java compilation failure!
```
[ERROR] .../mapping/test/AMapperImpl.java:[22,23] cannot find symbol
[ERROR] symbol:   variable c
[ERROR] location: class mapping.test.AMapperImpl
```

If you look at the generated mapping, it fails to define / instantiate the `c` variable that it later references:
```
public class AMapperImpl implements AMapper {

    @Override
    public APrime toAPrime(A a, APrime mappingTarget) {
        if ( a == null ) {
            return null;
        }

        if ( aBC( a ) != null ) {
            if ( mappingTarget.getcPrime() == null ) {
                mappingTarget.setcPrime( new CPrime() );
            }
            // ***** `c` was never defined!! *****
            // ***** Implementation should have done something like: *****
            // ***** `C c = aBC( a );` *****
            toCPrime( c, mappingTarget.getcPrime() );
        }
        else {
            mappingTarget.setcPrime( null );
        }

        return mappingTarget;
    }

    @Override
    public CPrime toCPrime(C c, CPrime mappingTarget) {
        if ( c == null ) {
            return null;
        }

        return mappingTarget;
    }

    private C aBC(A a) {
        if ( a == null ) {
            return null;
        }
        B b = a.getB();
        if ( b == null ) {
            return null;
        }
        C c = b.getC();
        if ( c == null ) {
            return null;
        }
        return c;
    }
}
```
Note: This has been tested with versions 1.2.0.Final and the latest beta 1.3.0.Beta2

**Comment by sjaakd** [comment]
Are you sure you both need update and creation behaviour? I'm considering whether we need support for this in the future. What's your use case?

**Comment by bcole** [comment]
Our use case is mapping from a DTO payload (`A`, `B`, `C`) to our JPA DB Entities (`APrime`, `CPrime`).

When a payload comes in to update an entity, we first find the existing JPA Entities (`APrime`, `CPrime`).  We need to pass these JPA Entities in as a `@MappingTarget` so that an update happens on the existing objects.  

If we don't pass `@MappingTarget`:
```
CPrime toCPrime(C c);
```
this would create a new CPrime JPA Entity (and generate a new ID), instead of using the existing entity.

**Comment by sjaakd** [comment]
I understand that you require update behaviour (so an ‘@MappingTarget‘ annotated parameter). What I don't understand is that you need a return type as well.. why not ‘void‘?

**Comment by bcole** [comment]
You're right, we don't need a return type.  Unfortunately, this doesn't solve the issue.

Here is the updated example Mapper:
```
@Mapper
public interface AMapper {
    @Mapping(target = "cPrime", source = "b.c")
    void toAPrime(A a, @MappingTarget APrime mappingTarget);

    void toCPrime(C c, @MappingTarget CPrime mappingTarget);
}
```

And the generated impl:
```
public class AMapperImpl implements AMapper {

    @Override
    public void toAPrime(A a, APrime mappingTarget) {
        if ( a == null ) {
            return;
        }

        if ( aBC( a ) != null ) {
            if ( mappingTarget.getcPrime() == null ) {
                mappingTarget.setcPrime( new CPrime() );
            }
            // ***** `c` is still not defined! *****
            toCPrime( c, mappingTarget.getcPrime() );
        }
        else {
            mappingTarget.setcPrime( null );
        }
    }

    ...

}

**Comment by sjaakd** [comment]
@bcole I'll investigate further.. Thanks for reporting!

**Comment by chris922** [comment]
@sjaakd I think I found the issue. At least the reproducer is working and all other tests are still green.. I will create a PR

**Comment by sjaakd** [comment]
solved in #1671

**Comment by sjaakd** [comment]
@bcole : no problem.. give it a shot.. its on the master..
